### ISSUE TITLE
Scope of the Project

### ISSUE THREAD

**Issue by Souvikns** [Tech-Phantoms/Auth#1]
### What is this Project?
I am starting this project because whatever we build will inevitably need authentication and as an organization, we should have our own authentication and user database. Thus this is what this project aims to solve. Auth will be a standalone service that will handle authentication and nothing else. It will have its own database where it will store all the user data that will be public and could be consumed by anyone. 

### Targeted Features. 
- Login and Signup UI with our ORGs logo. 
- public API to fetch registered users. (Just authenticate not Authorization. )

> @imabp should we make TechPhantoms OAUTH?

**Comment by imabp** [comment]
Yes I really want to make this up, we can even have a separate NPM Package, for tech phantoms auth only.. which we can use, to integrate with all our future projects.

**Comment by Souvikns** [comment]
So I would like to open up the discussion with some of the issues I am thinking about and how we can solve them. First, let me start with defining what `Auth` is aiming to serve. 

 Auth is a microservice that has only one job, to store and maintain users. It will have a database to store all the user data and an API to consume that. I and @imabp were thinking of using supabase for your authentication. Having GitHub OAuth would be really beneficial. So now that I define the aim let's talk about the issue. 

## Issues

### Using GitHub OAuth
Suppose we use GitHub OAuth then after the callback from GitHub it will redirect to our `Auth` URL, and not on any URL that the other service is listening on, then how will we use it for more than one service. As GitHub OAuth token is meant for one service only which this service will use, how can we have other services use this one service for their Authentication needs? 

I did come up with some solution let me discuss it here. 

- When other services are calling this service's URL to authenticate we store that data and after a complete authentication process redirects to that service's URL that could help. Something I and @imabp were talking about [you can check it here](https://stackoverflow.com/questions/17559050/how-to-pass-data-from-parent-html-window-to-pop-up-window). 
- Another approach is that we have separate GitHub OAuth tokens for all other Services but use this Service to store only the user data just making sure that we have the rights to push data but anyone can read some basic public data. 

### Having a perfect admin role system

I would like to have a service to support API_KEYS to allow others to create applications using this. So an admin system where anyone can first sign in then generates an API key to access the developer option. Now, this is something that is not required at the moment but is a feature that would be nice to have. Still, we need a way for us to use set up this service to build other services but that allows only us, not anyone else. 

---

> So @imabp @kalzen15 @syshcndr this is something we need to brainstorm and figure out. Lets keep discussing in this issue itself.

**Comment by imabp** [comment]
@Souvikns 
Reading through [GitHub OAuth](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps), I have a better design where we can approach this.
Using a single auth server to communicate between apps and oAuth.

Here is the flow [Download this as pdf](https://cdn.discordapp.com/attachments/849324748097716294/893752328421769256/authserver.pdf)
![image](https://user-images.githubusercontent.com/53480076/135706786-ec6ecf8c-b5f4-4282-83db-f6a53406d5fd.png)

**Comment by uzair-ali10** [comment]
@imabp @Souvikns 
Hello, This seems fascinating to me Can I also contribute to this??

**Comment by imabp** [comment]
Yes @uzair-ali10 why not. 
Its completely open source. In this issue, our main objective is to define and design the auth system. 
Please do share your inputs and improvements and once the auth system design is complete we will put it on code.

**Comment by Pranay-Narang** [comment]
> I would like to have a service to support API_KEYS to allow others to create applications using this. So an admin system where anyone can first sign in then generates an API key to access the developer option. Now, this is something that is not required at the moment but is a feature that would be nice to have. Still, we need a way for us to use set up this service to build other services but that allows only us, not anyone else.

This concept seems similar to generating **clients** in OAuth2.0 where each client is considered as an application and given a _client_id_ and _client_secret_ 

If I understood it right, you're essentially looking to build a service such as [Keycloak](https://www.keycloak.org/) which maintains a user database and supports third-party login such as GitHub right?. Please correct me if I am wrong

**Comment by imabp** [comment]
@Souvikns .. whats the status, are we going with the same.?

**Comment by imabp** [comment]
Create an epic if needed @Souvikns

**Comment by imabp** [comment]
Yes @Souvikns .. Let's create the issues and start working.
